### PR TITLE
Fixed #26967 -- Added support for AsGeoJSON, GeoHash and IsValid functions on MySQL.

### PR DESCRIPTION
**Pull request by sir-sigurd**
https://code.djangoproject.com/ticket/26967

### CODE REVIEW

File: docs/ref/contrib/gis/functions.txt [hunk 1]
```diff
@@ -56,7 +56,7 @@ geographic SRSes.
 
 .. class:: AsGeoJSON(expression, bbox=False, crs=False, precision=8, **extra)
 
-*Availability*: `PostGIS <https://postgis.net/docs/ST_AsGeoJSON.html>`__,
+*Availability*: MySQL, `PostGIS <https://postgis.net/docs/ST_AsGeoJSON.html>`__,
```
Comment (timgraham) [inline docs/ref/contrib/gis/functions.txt]: I think we're also including `(≥ 5.7.5)` here.

File: django/contrib/gis/db/models/functions.py [hunk 2]
```diff
@@ -322,6 +322,13 @@ def __init__(self, expression, precision=None, **extra):
             expressions.append(self._handle_param(precision, 'precision', int))
         super().__init__(*expressions, **extra)
 
+    def as_mysql(self, compiler, connection):
+        clone = self.copy()
+        # If no precision is provided, set it to maximum.
```
Comment (timgraham) [inline django/contrib/gis/db/models/functions.py]: the maximum

File: django/contrib/gis/db/backends/mysql/operations.py [hunk 3]
```diff
@@ -72,11 +72,13 @@ def function_names(self):
     @cached_property
     def unsupported_functions(self):
         unsupported = {
-            'AsGeoJSON', 'AsGML', 'AsKML', 'AsSVG', 'BoundingCircle',
-            'ForceRHR', 'GeoHash', 'IsValid', 'MakeValid', 'MemSize',
+            'AsGML', 'AsKML', 'AsSVG', 'BoundingCircle',
```
Comment (timgraham) [inline django/contrib/gis/db/backends/mysql/operations.py]: I'd be inclined to reflow this so it fits on three lines:
```
'AsGML', 'AsKML', 'AsSVG', 'BoundingCircle', 'ForceRHR',
'MakeValid', 'MemSize', 'Perimeter', 'PointOnSurface', 'Reverse',
'Scale', 'SnapToGrid', 'Transform', 'Translate',
```

File: docs/ref/contrib/gis/db-api.txt [hunk 4]
```diff
@@ -341,7 +341,7 @@ Lookup Type                        PostGIS    Oracle    MySQL [#]_   SpatiaLite
 :lookup:`equals`                   X          X         X            X          C
 :lookup:`exact`                    X          X         X            X          B
 :lookup:`intersects`               X          X         X            X          B
-:lookup:`isvalid`                  X          X                      X (LWGEOM)
+:lookup:`isvalid`                  X          X         X            X (LWGEOM)
```
Comment (timgraham) [inline docs/ref/contrib/gis/db-api.txt]: need to include version note here?

File: docs/ref/contrib/gis/functions.txt [hunk 5]
```diff
@@ -84,6 +84,12 @@ Keyword Argument       Description
                        representation -- the default value is 8.
 =====================  =====================================================
 
+On MySQL, ``crs`` parameter is futile.
```
Comment (timgraham) [inline docs/ref/contrib/gis/functions.txt]: I'd put this in the "Description" part of the table: "Ignored on MySQL."

File: tests/gis_tests/geoapp/test_functions.py [hunk 6]
```diff
@@ -220,9 +227,10 @@ def test_geohash(self):
         # SELECT ST_GeoHash(point) FROM geoapp_city WHERE name='Houston';
         # SELECT ST_GeoHash(point, 5) FROM geoapp_city WHERE name='Houston';
         ref_hash = '9vk1mfq8jx0c8e0386z6'
+
```
Comment (timgraham) [inline tests/gis_tests/geoapp/test_functions.py]: inadvertent addition?

File: tests/gis_tests/geoapp/test_functions.py [hunk 7]
```diff
@@ -58,9 +59,12 @@ def test_asgeojson(self):
         # SELECT ST_AsGeoJson("geoapp_city"."point", 8, 2) FROM "geoapp_city"
         # WHERE "geoapp_city"."name" = 'Houston';
         # This time we want to include the CRS by using the `crs` keyword.
+        if mysql:
+            houston_json = json.loads(houston_json)
```
Comment (timgraham) [inline tests/gis_tests/geoapp/test_functions.py]: Add a comment like: `# MySQL ignores the crs option.`?

File: tests/gis_tests/geoapp/test_functions.py [hunk 8]
```diff
@@ -58,9 +59,12 @@ def test_asgeojson(self):
...
+            del houston_json['crs']
         self.assertJSONEqual(
+            City.objects.annotate(json=functions.AsGeoJSON('point', crs=True)).get(name='Houston').json,
```
Comment (timgraham) [inline tests/gis_tests/geoapp/test_functions.py]: Is the assertion argument switching required as part of this patch or only cosmetic?

**Comment by timgraham** [comment]
buildbot, test on xenial.

**Comment by timgraham** [comment]
buildbot, test on xenial.

File: tests/gis_tests/geoapp/test_functions.py (hunk 8 above)
Comment (sir-sigurd) [inline tests/gis_tests/geoapp/test_functions.py]: It's required as `assertJSONEqual` accepts only strings as the first argument. The other issues are fixed. Thanks!

**Comment by sir-sigurd** [comment]
buildbot, test on xenial.

**Pull request merged**